### PR TITLE
Tutorials for derivs of explicit and implicit comps

### PR DESCRIPTION
**Pull request by JustinSGray**

### CODE REVIEW

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+
+****************************************************
+Defining Partial-Derivatives on Explicit Components
+****************************************************
+
+For any :ref:`ExplicitComponent <comp-type-2-explicitcomp>` instance you are going to provide derivatives of the **outputs with respect to the inputs**.
+Whenever you are going to define derivatives, there are two things you're required to do:
+
+    #. delclare the partial-derivatives via :code:`declare_partials`
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: typo: declare

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst [hunk 2]
```diff
@@ -0,0 +1,83 @@
+.. _advanced_guide_partial_derivs_explicit:
+
+OpenMDAO considers component derivatives to be **partial-derivatives**.
+The framework uses these partial-derivatives in order to compute the **total-derivatives** across your whole model.
+This tutorial is focused on how to define the partial-derivatives* for instances of :ref:`ExplicitComponent <comp-type-2-explicitcomp>`.
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: extraneous asterisk:  partial-derivatives*  (was this meant to be bolded)

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+
+
+****************************************************
+Defining Partial-Derivatives on Explicit Components
+****************************************************
+
+For any :ref:`ExplicitComponent <comp-type-2-explicitcomp>` instance you are going to provide derivatives of the **outputs with respect to the inputs**.
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: instance is not the right term here..  these are things you do in a class definition.. I think you could just remove the word instance

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+.. note::
+    There are a few more options to :code:`declare_partials` that are worth taking a look at.
+    There is support for when your derivatives are constant and if you want to specify derivatives in a sparse AIJ format.
+    The full details are :ref:`here <feature_specify_partials>`.
+
+After you declare the non-zero partial derivatives, you need to implement the :code:`compute_partials` method to perform the actual
+derivative computations.
+OpenMDAO will call this method whenever it needs to work with the partial derivatives.
+The values are stored in the Jacobian object, :code:`J`, and get used in the linear solutions that are necessary to compute model level total-derivatives.
+This API results in the actual assembly of the jacobian matrix in memory.
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: Suggest removing the word "actual".  
Also suggest moving these two sentences to the next paragraph.

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+It is the most appropriate way to declare derivatives in the vast majority of circumstances, and you should use it unless you have a good reason not to.
+
+Providing derivatives using the matrix-free API
+************************************************
+
+Sometimes you don't want to assemble the full partial-derivative Jacobian of your component in memory.
+We're not going to discuss why this might be the case in this tutorial.
+For now lets just say that if matrix-assembly won't work for your application you are likely already well aware of this issue.
+If you can't imagine why you would want to use a matrix-free API, then don't worry about this.
+If that is the case, there is matrix-free API you can use. See the :ref:`ExplicitComponent <comp-type-2-explicitcomp>` reference for an example of how to use it.
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: it looks like you inserted the previous line after writing this one, which changed what "that" means.   
I think "that" should mean "matrix-assembly won't work for you" but as written, it means "you can't imagine why you would want to use a matrix-free api."

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+
+How Do I Know if My Derivatives Are Correct?
+**************************************************
+
+Its really important, if you are going to provide analytic derivatives, that you make sure they are correct.
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: "It's" or "It is"

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst (hunk 2 above)
Comment (swryan) [reply openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: also I don't believe "partial derivatives" and  "total derivatives" should be hyphenated

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+.. embed-test::
+    openmdao.test_suite.test_examples.test_betz_limit.TestBetzLimit.test_betz_derivatives
+
+.. note::
+
+    :code:`check_partials` is **really** important when you're coding derivatives.
+    It has some options to give you more detailed outputs for debugging and to let you limit which components get tested.
+    You should look over the complete :ref:`check_partials doc <feature_check_partials>` before you start doing heavy development with derivatives.
+
+There is a lot of information there, including infromation about both forward and reverse derivatives.
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: typo: information

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+How Do I Know if My Derivatives Are Correct?
+**************************************************
+
+Its really important, if you are going to provide analytic derivatives, that you make sure they are correct.
+Nothing else will screw up the convergence of your analysis or optimization more quickly that bad derivatives.
+OpenMDAO provides a way to check your partial-derivatives so you can make sure they are right.
+Any time you implement analytic derivatives, or change the non-linear equations of your analysis you should check your partial-derivatives this way.
+
+.. embed-test::
+    openmdao.test_suite.test_examples.test_betz_limit.TestBetzLimit.test_betz_derivatives
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: note that the column headers don't line up in the output

File: openmdao/docs/advanced_guide/derivs/partial_derivs_implicit.rst [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+****************************************************
+Defining Partial-Derivatives on Implicit Components
+****************************************************
+
+For :ref:`ImplicitComponent <comp-type-3-implicitcomp>` instances you will provide partial-derivatives of **residuals with respect to inputs and outputs**.
+Note that this is slightly different than what you do for :ref:`ExplicitComponent instances <advanced_guide_partial_derivs_explicit>`, but
+the general procedure is similar:
+
+    #. delclare the partial-derivatives via :code:`declare_partials`
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_implicit.rst]: typo: declare

File: openmdao/docs/advanced_guide/derivs/partial_derivs_implicit.rst [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+provide derivatives of the :code:`apply_nonlinear` method to OpenMDAO.
+
+Here is a simple example to consider:
+
+.. embed-code::
+    openmdao.test_suite.components.quad_implicit.QuadraticComp
+
+
+In this component :code:`x` is an output and you take derivatives with respect to it.
+This might seem a bit strange to you, if you're used to thinking about things from an :ref:`ExplicitComponent instances <advanced_guide_partial_derivs_explicit>` perspective.
```
Comment (swryan) [inline openmdao/docs/advanced_guide/derivs/partial_derivs_implicit.rst]: suggest removing the comma and the word "instances"

File: openmdao/docs/advanced_guide/implicit_comps/implicit_with_balancecomp.rst [hunk 11]
```diff
@@ -0,0 +1,72 @@
+In the :ref:`previous tutorial <defining_icomps_tutorial>`, we built up a model of an electrical circuit using
+a combination of :ref:`ImplicitComponent <comp-type-3-implicitcomp>` and :ref:`ExplicitComponent <comp-type-2-explicitcomp>` instances.
+In that tutorial, all of the implicit relationships in the model came directly from the physics of the model itself.
+
+However, you often need to add implicit relatiships to models by driving the values of two separate variables to be equal to eachother.
```
Comment (swryan) [inline openmdao/docs/advanced_guide/implicit_comps/implicit_with_balancecomp.rst]: typo "each other" (missing space)

File: openmdao/docs/advanced_guide/implicit_comps/implicit_with_balancecomp.rst [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+But OpenMDAO provides :ref:`BalanceComp <balancecomp_feature>`, a general utility component, designed specifically for this type of situation.
+
+What we're going to do is add a :code:`BalanceComp` to the top level of the model that will drive the input current to the circuit to force a delta-V across the battery to be what we want.
+We'll also add an :ref:`ExecComp <feature_exec_comp>` to compute that delta-V from the ground voltage and the voltage at node 1 and then connect everything up.
+Lastly, since we added an :ref:`ImplicitComponent <comp-type-3-implicitcomp>` at the top level of the model, we'll also move the :ref:`NewtonSolver <nlnewton>` up to the top level of the model too.
+
+.. note::
+
+    BalanceComp can handle more than just :math:`lhs-rhs=0`. It has a number of inputs that let you tweak that behavior.
+    It can support multiple residuals, and array variables as well. Check out the :ref:`feature doc <balancecomp_feature>` on it for details.
```
Comment (swryan) [inline openmdao/docs/advanced_guide/implicit_comps/implicit_with_balancecomp.rst]: I think "feature doc" is inside baseball.. suggest just "documentation"

File: openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst (hunk 8 above)
Comment (JustinSGray) [reply openmdao/docs/advanced_guide/derivs/partial_derivs_explicit.rst]: Yeah. I put up a bug story for this. There is a problem with the embed directive not properly html escaping `<` and `>` which show up in the output txt

File: openmdao/docs/advanced_guide/implicit_comps/implicit_with_balancecomp.rst [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+.. math::
+    \mathcal{R}_{batt} = V1 - V0 - V_{source}^{*}
+
+where :math:`V_{source}^{*}`, the desired source voltage, is given by the user as parameter to the model.
+
+We could write a new component, inheriting from :ref:`ImplicitComponent <comp-type-3-implicitcomp>`, to include this new relationship into the model.
+But OpenMDAO provides :ref:`BalanceComp <balancecomp_feature>`, a general utility component, designed specifically for this type of situation.
+
+What we're going to do is add a :code:`BalanceComp` to the top level of the model that will drive the input current to the circuit to force a delta-V across the battery to be what we want.
```
Comment (kmarsteller) [inline openmdao/docs/advanced_guide/implicit_comps/implicit_with_balancecomp.rst]: This sentence is a little run-on...  Maybe put a period after "model." Then start, "The BalanceComp will drive the..." ?  Author's choice if you don't want to.

File: openmdao/docs/advanced_guide/implicit_comps/implicit_with_balancecomp.rst [hunk 14]
```diff
@@ -18,7 +18,7 @@ circuit with a current source of .1 Amps. Here is a reminder of what the circuit
    :alt: diagram of a simple circuit with two resistors and one diode
 
 When we solved that circuit, the resulting voltage at *node 1* was 9.9 Volts.
-Lets say you wanted to power this circuit off a 1.5 Volt battery, instead of a current source.
+Lets say you wanted to power this circuit with a 1.5-Volt battery, instead of a current source.
```
Comment (kmarsteller) [inline openmdao/docs/advanced_guide/implicit_comps/implicit_with_balancecomp.rst]: Still need an apostrophe in "Let's" (sorry)

**Pull request merged**